lib/smack/SmackInstGenerator.cpp

```diff
@@ -144,6 +144,37 @@ void SmackInstGenerator::processInstruction(llvm::Instruction& inst) {
   nextInst++;
 }
 
+void SmackInstGenerator::processIntrinsicCall(llvm::IntrinsicInst* ii) {
+  static auto f16Cast = [&ii, this](bool f16UpCast) {
+    if (SmackOptions::BitPrecise && SmackOptions::FloatEnabled) {
+      std::string funcName;
+      auto tyID = f16UpCast?
+        ii->getFunctionType()->getReturnType()->getTypeID():
+        ii->getArgOperand(0)->getType()->getTypeID();
+      if (tyID == llvm::Type::FloatTyID)
+        funcName = f16UpCast? "__SMACK_convert_from_fp16_f32":
+          "__SMACK_convert_to_fp16_f32";
+      else if (tyID == llvm::Type::DoubleTyID)
+        funcName = f16UpCast? "__SMACK_convert_from_fp16_f64":
+          "__SMACK_convert_to_fp16_f64";
+      return Stmt::call(funcName, {rep->expr(ii->getArgOperand(0))},
+            {naming->get(*ii)});
+    } else
+      return rep->call(ii->getCalledFunction(), *ii);
+  };
+  static std::map<llvm::Intrinsic::ID, std::function<const Stmt*()>> stmtMap {
+    {llvm::Intrinsic::convert_from_fp16,
+      []() { return f16Cast(true); }},
```

Comment:
Instead of passing in true/false to signal up/down cast, and then branching on this in `f16Cast`, why wouldn't you instead split f16Cast into two function, one for up and one for down casting?
That would make the code cleaner since you would avoid branching on `f16UpCast` boolean.

Reply:
Having two functions was our previous implementation then I realized that the two functions share a lot of code like `if (SmackOptions::...)`. I think the code is actually cleaner with this switch.

Reply:
Actually, what has also been bugging me is the fact that we do not explicitly specify models for these intrinsics for the cases where `--bit-precise` and `--floats` is not set. I find the fact that these models are implicitly generated confusing.
So I propose that we include those models as well in the `intrinsics.c` file, in which case `if (SmackOptions::...)` conditions would disappear.